cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -0,0 +1,880 @@
+           0 == memcmp(value, dict_data + dict_offsets[slot], length))) {           \
+    ++j;                                                                            \
+    if (ARROW_PREDICT_FALSE(j == hash_table_size_)) {                               \
+      j = 0;                                                                        \
+    }                                                                               \
+    slot = hash_slots_[j];                                                          \
+  }                                                                                 \
+                                                                                    \
+  if (slot == kHashSlotEmpty) {                                                     \
+    if (!Action::allow_expand) {                                                    \
+      throw HashException("Encountered new dictionary value");                      \
+    }                                                                               \
+                                                                                    \
+    slot = dict_size_++;                                                            \
+    hash_slots_[j] = slot;                                                          \
+                                                                                    \
+    RETURN_NOT_OK(dict_data_.Append(value, length));                                \
+    RETURN_NOT_OK(dict_offsets_.Append(static_cast<int32_t>(dict_data_.length()))); \
+                                                                                    \
+    action->ObserveNotFound(slot);                                                  \
+                                                                                    \
+    if (ARROW_PREDICT_FALSE(dict_size_ > hash_table_size_ * kMaxHashTableLoad)) {   \
+      RETURN_NOT_OK(action->DoubleSize());                                          \
+    }                                                                               \
+  } else {                                                                          \
+    action->ObserveFound(slot);                                                     \
+  }
+
+    GENERIC_HASH_PASS(HASH_INNER_LOOP);
+
+#undef HASH_INNER_LOOP
+
+    return Status::OK();
+  }
+
+  Status GetDictionary(std::shared_ptr<ArrayData>* out) override {
+    // TODO(wesm): handle null being in the dictionary
+    BufferVector buffers = {nullptr, nullptr, nullptr};
+
+    RETURN_NOT_OK(dict_offsets_.Finish(&buffers[1]));
+    RETURN_NOT_OK(dict_data_.Finish(&buffers[2]));
+
+    *out = std::make_shared<ArrayData>(type_, dict_size_, std::move(buffers), 0);
+    return Status::OK();
+  }
+
+ protected:
+  int64_t HashValue(const uint8_t* data, int32_t length) const {
+    return HashUtil::Hash(data, length, 0);
+  }
+
+  Status DoubleTableSize() {
```

Comment:
This is duplicated between the different HashKernel implementations?

Reply:
The hot path on the inner loop of the hash table is different because of the different memory layout (primitive vs. varbytes vs fixed length bytes). 

Getting rid of the code duplication leaves some unsavory options:

* Writing more macros (debugging harder)
* Introducing some auxiliary structs / functions to make the non-primitive arrays "value-like". E.g. in the old code there was the `WrappedBinary` type. We obviously need to write some benchmarks, but I suspect this would make things slower

I think we should write a bunch of microbenchmarks for various cases (small cardinality categories, large cardinality categories) and use this implementation as the performance baseline. If we refactor to add different layers of indirection to enable code reuse and the performance gets worse, we should not make such changes.

FWIW, I am OK with writing large amounts of nested macros in this particular set of kernels, particularly since the hot paths in this code, with the exception of appending to the dictionary, cannot fail. It's also useful to be able to easily change the hash table probing stretch from linear to quadratic by switching the macro name used

Reply:
Added macros. In this particular case, I don't think it's so bad